pilot/pkg/model/push_context.go

```diff
@@ -320,11 +326,41 @@ func (ps *PushContext) initVirtualServices(env *Environment) error {
 		return err
 	}
 
+	type combinedVirtualService struct {
+		// What are the parameters to order?
+		// Virtual services have multiple hosts and multiple protocols. We need to split it into
+		// a multiple virtual services each with one host and one protocol type http/tcp/tls.
+		// then we need to look at the set of all virtual services for the same host/protocol
+		// and try to merge these services into one
+		//   Whats the ordering algorithm?
+		//    For http, sort all the match conditions for the virtual services for the host
+		//     specific port before no port
+		//     specific gateway before no gateway
+		//     specific source label before no source label
+		//     uri match before no uri match
+		//         between two URIs,
+		//           exact before prefix
+		//           prefix before regex
+		//           longer value (exact/prefix/regex) before the shorter one
+		//     authority match before no authority match
+		//     method before no method
+		//     specific scheme before no scheme
+		//     headers before no headers
+		//     all else being equal, tie break by time
```

Comment:
Hmm, do we also need to ensure that all of the clauses are in the same relative order as their original VS (i.e. can I interleave clauses from separate VSs so long as, when I look at the clauses from a single VS, they're in the same order as the original document)? Certainly for regexes we need to do this. It's unclear to me if there are cases in prefix match clauses where order needs to be preserved that wouldn't already be covered by the "longest to shortest" sorting.